Crane/app/src/main/java/androidx/compose/samples/crane/details/DetailsActivity.kt

```diff
@@ -177,58 +173,69 @@ fun DetailsContent(
 }
 
 @Composable
-private fun CityMapView(latitude: String, longitude: String) {
-    // The MapView lifecycle is handled by this composable. As the MapView also needs to be updated
-    // with input from Compose UI, those updates are encapsulated into the MapViewContainer
-    // composable. In this way, when an update to the MapView happens, this composable won't
-    // recompose and the MapView won't need to be recreated.
-    val mapView = rememberMapViewWithLifecycle()
-    MapViewContainer(mapView, latitude, longitude)
-}
-
-@Composable
-private fun MapViewContainer(
-    map: MapView,
+fun CityMapView(
     latitude: String,
-    longitude: String
+    longitude: String,
+    cameraState: CameraPositionState = rememberSaveable(saver = CameraPositionState.Saver) { CameraPositionState() },
+    onMapLoaded: () -> Unit = {}
 ) {
-    val cameraPosition = remember(latitude, longitude) {
+    val cityLocation = remember(latitude, longitude) {
         LatLng(latitude.toDouble(), longitude.toDouble())
     }
 
-    LaunchedEffect(map) {
-        val googleMap = map.awaitMap()
-        googleMap.addMarker { position(cameraPosition) }
-        googleMap.moveCamera(CameraUpdateFactory.newLatLng(cameraPosition))
+    cameraState.position = CameraPosition.fromLatLngZoom(
+        cityLocation,
+        InitialZoom
+    )
+
+    MapViewContainer(
+        cameraPositionState = cameraState,
+        onMapLoaded = onMapLoaded
+    ) {
+        Marker(position = cityLocation)
     }
+}
 
-    var zoom by rememberSaveable(map) { mutableStateOf(InitialZoom) }
-    ZoomControls(zoom) {
-        zoom = it.coerceIn(MinZoom, MaxZoom)
+@Composable
+private fun MapViewContainer(
+    cameraPositionState: CameraPositionState,
+    onMapLoaded: () -> Unit = {},
+    content: (@Composable () -> Unit)? = null
+) {
+    val mapProperties = remember {
+        MapProperties(
+            maxZoomPreference = MaxZoom,
+            minZoomPreference = MinZoom,
+        )
     }
 
-    val coroutineScope = rememberCoroutineScope()
-    AndroidView({ map }) { mapView ->
-        // Reading zoom so that AndroidView recomposes when it changes. The getMapAsync lambda
-        // is stored for later, Compose doesn't recognize state reads
-        val mapZoom = zoom
-        coroutineScope.launch {
-            val googleMap = mapView.awaitMap()
-            googleMap.setZoom(mapZoom)
-            // Move camera to the same place to trigger the zoom update
-            googleMap.moveCamera(CameraUpdateFactory.newLatLng(cameraPosition))
-        }
+    val mapUiSettings = remember {
+        // We are providing our own zoom controls so disable the built-in ones.
+        MapUiSettings(zoomControlsEnabled = false)
     }
+
+    ZoomControls(
```

Comment:
Should `ZoomControls` and `GoogleMap` be wrapped in a `Column`? That'd make it more reusable as right now, it works because higher up in the hierarchy, there's a `Column`